### PR TITLE
Expose HTTP request information in response

### LINKED ISSUE #440: Expose Request information on Response
In most HTTP request libraries a `response` has a `request` property. This is useful for debugging, logging, generating reports etc.
In k6 this would be very useful for usage with `check`. 

I would propose the following approach:

1. Create an `HTTPRequest` struct.
2. Instantiate this struct in `HTTP.Request` based on all args.
3. Add a `Request` property on the `HTTPResponse` struct.

### PR DESCRIPTION
**Pull request by cstyan**
Fixes #440 

Adds HTTPRequest struct, includes the request as a field in the response, and finally a test for whether or not the request information is properly populated in a response.

cc: @robingustafsson @marklagendijk

### CODE REVIEW

**Comment by marklagendijk** [comment]
Well done! LGTM!

File: js/modules/k6/http/http_request.go [hunk 1]
```diff
@@ -109,9 +117,14 @@ func (h *HTTP) request(ctx context.Context, rt *goja.Runtime, state *common.Stat
 		URL:    url.URL,
 		Header: make(http.Header),
 	}
+	HTTPReq := &HTTPRequest{
```
Comment (robingustafsson) [inline js/modules/k6/http/http_request.go]: Could we change the name to `respReq` or something more descriptive (and start with lowercase letter to mainntain similar var name style)?

File: js/modules/k6/http/http_request.go [hunk 2]
```diff
@@ -109,9 +117,14 @@ func (h *HTTP) request(ctx context.Context, rt *goja.Runtime, state *common.Stat
...
+		Method: req.Method,
+		URL:    req.URL,
```
Comment (robingustafsson) [inline js/modules/k6/http/http_request.go]: Hm, I think this should be `req.URL.URLString` to be consistent with `res.url` (which is just a string). We should expose URL parsing/joining/composing functionality in a specific module though.

File: js/modules/k6/http/response.go [hunk 3]
```diff
@@ -62,6 +62,8 @@ type HTTPResponse struct {
 	OCSP           OCSP `js:"ocsp"`
 	Error          string
```
Comment (robingustafsson) [inline js/modules/k6/http/response.go]: You can remove the newline here.

File: js/modules/k6/http/http_request.go [hunk 4]
```diff
@@ -44,6 +44,14 @@ import (
 	null "gopkg.in/guregu/null.v3"
 )
 
+type HTTPRequest struct {
+	Method  string
+	URL     *neturl.URL
+	Headers map[string][]string
+	Body    io.Closer
+	Cookies map[string]*HTTPRequestCookie
```
Comment (robingustafsson) [inline js/modules/k6/http/http_request.go]: The `Cookie` header is allowed to contain multiple cookies with the same name [1] so `Cookies` here needs to be a `map[string][]*HTTPRequestCookie`, or `map[string][]string` would be fine as well in my opinion.

[1] - https://tools.ietf.org/html/rfc6265#section-4.2

File: js/modules/k6/http/http_request.go [hunk 5]
```diff
@@ -235,7 +248,10 @@ func (h *HTTP) request(ctx context.Context, rt *goja.Runtime, state *common.Stat
 		}
 	}
 
-	resp := &HTTPResponse{ctx: ctx, URL: url.URLString}
+	HTTPReq.Headers = req.Header
+	HTTPReq.Cookies = reqCookies
```
Comment (robingustafsson) [inline js/modules/k6/http/http_request.go]: The cookies sent with the request could also be retrieved from the active cookiejar so we need similar logic to what is in https://github.com/loadimpact/k6/blob/master/js/modules/k6/http/http.go#L130 (called a right above this) to determine the actual cookies sent. I guess the easiest would be to have `setRequestCookies()` return the determined set of cookies to send (`jarCookies`).

File: js/modules/k6/http/http_request.go (hunk 4 above)
Comment (cstyan) [reply js/modules/k6/http/http_request.go]: Here I've just used the same types in the map as are being used for the cookies being put into the request struct: https://github.com/loadimpact/k6/blob/master/js/modules/k6/http/http_request.go#L139 

Does that map need to change as well?

File: js/modules/k6/http/http_request.go (hunk 1 above)
Comment (cstyan) [reply js/modules/k6/http/http_request.go]: :+1:

File: js/modules/k6/http/http_request.go (hunk 5 above)
Comment (cstyan) [reply js/modules/k6/http/http_request.go]: Not sure I understand completely. I think you're saying we want to add `getRequestCookies()` to get the cookies that would have actually have been put into the `http.Request` struct?

File: js/modules/k6/http/http_request.go (hunk 4 above)
Comment (robingustafsson) [reply js/modules/k6/http/http_request.go]: No, that map only stores the user specified per-request cookies (from JS, like the first example here: https://docs.k6.io/docs/cookies#section-setting-simple-cookies).

But to get all cookies that are sent with the request (we is what we want to expose through `res.request.cookies`) we also need to look at what's in the cookiejar (from any previous interactions with the server). And that's when the situation can arise where we end up with multiple cookies with the same name.

Merging the per-request cookies set by the user in JS land with any cookies already present in the cookiejar is what https://github.com/loadimpact/k6/blob/master/js/modules/k6/http/http.go#L130 does.

**Comment by cstyan** [comment]
@robingustafsson I've changed the cookies to `[]*http.Cookie`, which is the same type returned by `http.Request.Cookies()`. I didn't see a nice way without some more refactoring to have `[]*http.Cookie` in the actual request and `map[string][]*HTTPRequestCookie` or `map[string][]string` in the request struct that is returned via the response.

**Comment by marklagendijk** [comment]
@cstyan can't you just do something like this?
``` go
cookies := req.Cookies()
respReq.Cookies := make(map[string][]string)
for _, cookie := range cookies {
	if array, exists := respReq.Cookies[cookie.Name]; exists{
		respReq.Cookies[cookie.Name] = append(array, cookie.Value)
	} else {
		respReq.Cookies[cookie.Name] = []string { cookie.Value }
	}
	
}
```

File: js/modules/k6/http/http_request.go [hunk 6]
```diff
@@ -25,6 +25,7 @@ import (
 	"compress/gzip"
 	"compress/zlib"
 	"context"
+	// "fmt"
```
Comment (marklagendijk) [inline js/modules/k6/http/http_request.go]: Can be removed.

File: js/modules/k6/http/http_request_test.go [hunk 7]
```diff
@@ -847,4 +847,27 @@ func TestRequestAndBatch(t *testing.T) {
...
+			let jar = new http.CookieJar();
+
+			jar.set("https://httpbin.org/cookies", "key", "value");
+			res = http.request("GET", "https://httpbin.org/cookies", null, { cookies: { key2: "value2" }, jar: jar });
+
+			if (res.json().cookies.key != "value") { throw new Error("wrong cookie value: " + res.json().cookies.key); }
+
+			if (res.status != 200) { throw new Error("wrong status: " + res.status); }
+			if (res.request["method"] !== "GET") { throw new Error("http request method was not \"GET\": " + JSON.stringify(res.request)) }
+			if (res.request["body"] != null) { throw new Error("http request body was not null: " + JSON.stringify(res.request)) }
```
Comment (marklagendijk) [inline js/modules/k6/http/http_request_test.go]: Missing test case for when body should be set. When testing this myself it seems that `request.body` does not work:
```
panic: reflect.Value.Addr of unaddressable value [recovered]--------] 0 / 1
        panic: Panic at 86: reflect.Value.Addr of unaddressable value [recovered]
        panic: Panic at 86: reflect.Value.Addr of unaddressable value

goroutine 37 [running]:
github.com/loadimpact/k6/vendor/github.com/dop251/goja.AssertFunction.func1.1(0xc422d2bd00)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/runtime.go:1407 +0xa2
panic(0xb53720, 0xc4201360d0)
        /home/markl/.gvm/gos/go1.9.2/src/runtime/panic.go:491 +0x283
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).try.func1(0xc42203b450, 0x0, 0xc422d2bc08, 0x0, 0x0, 0x0, 0xc422d2bc60)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:364 +0x4b4
panic(0xb2b620, 0xcdc1d0)
        /home/markl/.gvm/gos/go1.9.2/src/runtime/panic.go:491 +0x283
reflect.Value.Addr(0xb70f00, 0xc420609a30, 0x94, 0xc423036280, 0x1, 0x1)
        /home/markl/.gvm/gos/go1.9.2/src/reflect/value.go:239 +0x82
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*objectGoReflect)._getField(0xc42226e510, 0xc423036268, 0x6, 0x411eb7, 0xc4201360a0, 0x10)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/object_goreflect.go:93 +0xf8
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*objectGoReflect)._get(0xc42226e510, 0xc423036268, 0x6, 0x10, 0xc4201360a0)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/object_goreflect.go:111 +0xfc
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*objectGoReflect).getStr(0xc42226e510, 0xc423036268, 0x6, 0xc423055e20, 0x20)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/object_goreflect.go:124 +0x43
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*objectGoReflect).get(0xc42226e510, 0x1131f00, 0xc420136030, 0x18, 0xc423055e20)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/object_goreflect.go:86 +0x3f
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*_builtinJSON_stringifyContext).str(0xc421b8c820, 0x1131f00, 0xc420136030, 0xc4230551c0, 0x0)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/builtin_json.go:283 +0x6e
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*_builtinJSON_stringifyContext).jo(0xc421b8c820, 0xc4230551c0)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/builtin_json.go:458 +0x1f1
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*_builtinJSON_stringifyContext).str(0xc421b8c820, 0x1131f00, 0xc4203b9d20, 0xc423054900, 0x0)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/builtin_json.go:370 +0x3be
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*_builtinJSON_stringifyContext).jo(0xc421b8c820, 0xc423054900)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/builtin_json.go:458 +0x1f1
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*_builtinJSON_stringifyContext).str(0xc421b8c820, 0x1131f00, 0x12bfc00, 0xc4230556c0, 0x12bfc00)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/builtin_json.go:370 +0x3be
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*_builtinJSON_stringifyContext).do(0xc421b8c820, 0x1131d20, 0xc423054900, 0xc423036334)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/builtin_json.go:279 +0x1de
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*Runtime).builtinJSON_stringify(0xc422870480, 0x1131d20, 0xc42187ab80, 0xc423121e60, 0x3, 0x3, 0xc423055640, 0xc420c18500)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/builtin_json.go:270 +0x7d4
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*Runtime).(github.com/loadimpact/k6/vendor/github.com/dop251/goja.builtinJSON_stringify)-fm(0x1131d20, 0xc42187ab80, 0xc423121e60, 0x3, 0x3, 0x3, 0x4126f0)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/builtin_json.go:517 +0x51
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*Runtime).functionproto_apply(0xc422870480, 0x1131d20, 0xc42187abe0, 0xc421c9d450, 0x2, 0x39, 0xc423036310, 0xc422d2b8c0)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/builtin_function.go:59 +0x143
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*Runtime).(github.com/loadimpact/k6/vendor/github.com/dop251/goja.functionproto_apply)-fm(0x1131d20, 0xc42187abe0, 0xc421c9d450, 0x2, 0x39, 0x5, 0x1)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/builtin_function.go:159 +0x51
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm)._nativeCall(0xc42203b450, 0xc423004370, 0x2)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:1826 +0x2ec
github.com/loadimpact/k6/vendor/github.com/dop251/goja.call.exec(0x2, 0xc42203b450)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:1810 +0x6b3
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*call).exec(0xc42283c0ec, 0xc42203b450)
        <autogenerated>:1 +0x44
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).run(0xc42203b450)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:288 +0x51
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*funcObject).Call(0xc420ca1500, 0x1132360, 0x12e2100, 0x0, 0x0, 0x0, 0x456c17, 0xc42002c600)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/func.go:130 +0x3c8
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*funcObject).Call-fm(0x1132360, 0x12e2100, 0x0, 0x0, 0x0, 0x42a5c9, 0xc4212c0030)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/builtin_function.go:120 +0x51
github.com/loadimpact/k6/vendor/github.com/dop251/goja.AssertFunction.func1.2()
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/runtime.go:1412 +0xb6
github.com/loadimpact/k6/vendor/github.com/dop251/goja.(*vm).try(0xc42203b450, 0xc420049c78, 0x0)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/vm.go:370 +0x119
github.com/loadimpact/k6/vendor/github.com/dop251/goja.AssertFunction.func1(0x1132360, 0x12e2100, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/vendor/github.com/dop251/goja/runtime.go:1411 +0x12e
github.com/loadimpact/k6/js.(*VU).RunOnce(0xc4216ac930, 0x112d240, 0xc4216d00c0, 0x0, 0xc4222d4bb0, 0xc420b979f0, 0xc420b95bb0, 0xc420b979e0)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/js/runner.go:260 +0x4c1
github.com/loadimpact/k6/core/local.(*vuHandle).run(0xc421517180, 0xc42007aaa0, 0xc42310e1e0, 0xc42310e180)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/core/local/local.go:64 +0x1cc
github.com/loadimpact/k6/core/local.(*Executor).scale.func1(0xc421517180, 0xc423177c20, 0xc42310e1e0, 0xc42310e180)
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/core/local/local.go:315 +0x4d
created by github.com/loadimpact/k6/core/local.(*Executor).scale
        /home/markl/.gvm/pkgsets/go1.9.2/global/src/github.com/loadimpact/k6/core/local/local.go:314 +0x35f
```

Comment (cstyan) [reply js/modules/k6/http/http_request_test.go]: oof, okay I'll look into this!

**Comment by cstyan** [comment]
I think this is good to go

**Comment by robingustafsson** [comment]
@cstyan Yes, looks good. The only thing I'm a bit concerned about it is the added memory copy resulting from `respReq.Body = bodyBuf.String()`. In the upcoming Go 1.10 release there's a new `strings.Builder` that we could perhaps make use of later  (https://github.com/golang/go/commit/37b056948d058679efa4e87fb6c9b2a2ddfa31a3) to avoid the memory copy, but we'll have to look closer at this for https://github.com/loadimpact/k6/pull/370 as well where  `respReq.Body = bodyBuf.String()` will probably change to `respReq.Body = bodyBuf.Bytes()`, that would afaict also avoid a memory copy.

I'll merge this PR and then evaluate how the memory copy can be avoided as part of https://github.com/loadimpact/k6/pull/370 (as it's mostly going to be costly when doing larger uploads, binary data like images, videos etc.).

**Pull request merged**